lib/remote/endpoint.cpp

```diff
@@ -38,11 +38,49 @@ void Endpoint::OnAllConfigLoaded()
 {
 	ObjectImpl<Endpoint>::OnAllConfigLoaded();
 
+	String zoneName = GetZoneName();
+	if (zoneName != "") {
+		Zone::Ptr zone = Zone::GetByName(zoneName);
+		if (!zone) {
+			BOOST_THROW_EXCEPTION(ScriptError("Endpoint '" + GetName() +
+				"' cannot belong to zone '" + zoneName +
+				"', zone does not exist.", GetDebugInfo()));
+		}
+
+		zone->AddEndpoint(this);
+
+		// Adjust runtime zone attr value so that object is treated appropriately.
+		Zone::Ptr parent = zone->GetParent();
+		if (parent) {
+			Log(LogDebug, "Endpoint")
+				<< "Update endpoint '" << GetName() << "' zone attribute to '" << parent->GetName() <<
+				"' (retain endpoint object sync).";
+			SetZoneName(parent->GetName());
+		}
+
+		// XXX: If/When AddConnection() becomes public
+		// Attempt to connect to endpoint.
+		//Zone::Ptr local = Zone::GetLocalZone();
+		//if (parent == local) {
+		//	ApiListener::Ptr listener = ApiListener::GetInstance();
+		//	if (listener)
+		//		listener->AddConnection(this);
+		//}
+	}
+
 	if (!m_Zone)
 		BOOST_THROW_EXCEPTION(ScriptError("Endpoint '" + GetName() +
 			"' does not belong to a zone.", GetDebugInfo()));
```

Comment:
If at all(!)\*, I personally would relax this check instead. This way, one could prepare some endpoints and finally assemble them to a zone.

* \*) https://github.com/Icinga/icinga2/issues/3823#issuecomment-1268152602